testcontainers/src/core/containers/request.rs

```diff
@@ -49,6 +49,64 @@ pub struct ContainerRequest<I: Image> {
+    /// CFS will schedule the threads for this container. Normally you don't adjust this and
+    /// just set the CPU quota or nano CPUs. You might want to set this if you want to increase
+    /// or reduce context-switching the container is subjected to.
+    pub(crate) cpu_period: Option<i64>,
+
+    /// Microseconds of CPU time that the container can get in a CPU period.
+    /// Most users will want to set CPU quota to their desired CPU count * 100000.
+    /// For example, to limit a container to 2 CPUs, set CPU quota to 200000.
+    /// This is based on the default CPU period of 100000.
+    /// If CPU quota is set to 0, the container will not be limited.
+    pub(crate) cpu_quota: Option<i64>,
+
+    /// The length of a CPU real-time period in microseconds. Set to 0 to allocate no time allocated to real-time tasks.
+    pub(crate) cpu_realtime_period: Option<i64>,
+
+    /// The length of a CPU real-time runtime in microseconds. Set to 0 to allocate no time allocated to real-time tasks.
+    pub(crate) cpu_realtime_runtime: Option<i64>,
+
+    /// CPUs in which to allow execution (e.g., `0-3`, `0,1`).
+    /// Core pinning should help with performance consistency and context switching in some cases.
+    pub(crate) cpuset_cpus: Option<String>,
+
+    /// CPU quota in units of 10<sup>-9</sup> CPUs. This is basically what the --cpus flag turns into, but the
+    /// raw value is denominated in billionths of a CPU. cpu_period and cpu_quota give you more control over the scheduler.
+    pub nano_cpus: Option<i64>,
+
+    /// Memory limit for the container, the _minimum_ is 6 MiB.
+    /// This is the same as `HostConfig::memory`.
+    pub(crate) memory: Option<i64>,
+
+    /// Memory reservation, soft limit. Analogous to the JVM's `-Xms` option.
+    /// The _minimum_ is 6 MiB.
+    /// This is the same as `HostConfig::memory_reservation`.
+    pub(crate) memory_reservation: Option<i64>,
+
+    /// Total memory limit (memory + swap). Set as `-1` to enable unlimited swap.
+    /// Same 6 MiB minimum as `memory`.
+    pub memory_swap: Option<i64>,
+
+    /// Tune a container's memory swappiness behavior. Accepts an integer between 0 and 100.
+    pub memory_swappiness: Option<i64>,
+
+    /// Disable OOM Killer for the container. This will not do anything unless -m (memory limit, cf. memory on this struct) is set.
+    /// You can disable OOM-killer by writing "1" to memory.oom_control file, as:
+    /// ```ignore
+    /// echo 1 > memory.oom_control
+    /// ```
+    /// This operation is only allowed to the top cgroup of sub-hierarchy.
+    /// If OOM-killer is disabled, tasks under cgroup will hang/sleep
+    /// in memory cgroup's OOM-waitqueue when they request accountable memory.
+    /// https://lwn.net/Articles/432224/
+    pub oom_kill_disable: Option<bool>,
+
+    /// Tune a container's PIDs limit. Set `0` or `-1` for unlimited, or `null` to not change.
+    pub pids_limit: Option<i64>,
```

Comment:
That's quite a lot of options
It feels like it would be nice group some of them to structs (e.g `CpuLimits`, `MemoryLimits` - naming is questionable though) and accept typed arguments within `ImageExt`

This would provide us more granularity and the same time less overloaded trait

Reply:
Testcontainers Go maintainer here. Instead of adding so many options, we usually prefer passing a function modifier that has access to the different internal docker types, so the user can, for advanced cases, change all of them if needed: https://golang.testcontainers.org/features/creating_container/#advanced-settings

Please take a look at the config, host config and endpoint settings modifiers.

Of course, I'm not sure how mature is the Rust docker client nowadays.

Reply:
I was following the pattern of the existing Rust API. If the Rust maintainers want to refactor that, that's fine, but it's out of scope here. I've been using this fork for the better part of a year because I needed control over these parameters.

Reply:
Agreed. That will enable more advanced use cases without increasing the API surface that much

Reply:
I'd really like to unfork my dependency. If you want to change the API you're free to do that at any time.

Reply:
I will try to dedicate time and implement it If you don't have capacity to perform the requested changes here.